### DISCUSSION TITLE
"Running with systemd (docker)" don't launch in the background and don't restart

### DISCUSSION THREAD

**Discussion by ctschach** [dani-garcia/vaultwarden#2026] [Q&A]
### Subject of the issue

"Running with systemd (docker)" don't launch in the background and don't restart

I'm starting the docker instance of Vaultwarden via systemd as described in this Wiki:

https://github.com/dani-garcia/vaultwarden/wiki/Running-with-systemd-docker

However, I noticed two problem:

* "systemctl start vaultwarden.service" does not run in the background. It launches the docker instance, but never come back to the shell. So I had to use Ctrl-Z and "bg" to let it run in the background.
* At some point the docker process was not running anymore. Not sure what the problem was, but it did not automatically restart - which should happen when you use systemd.

### Deployment environment
* OS:
Ubuntu 18.04LTS on x86

* vaultwarden version:
vaultwarden/server:latest

* Install method:
docker.

### Configuration

Here is the content of my service-file:
/etc/systemd/system/vaultwarden.service

```
[Unit]
Description=Vaultwarden
After=docker.service
Requires=docker.service

[Service]
TimeoutStartSec=0
ExecStartPre=-/usr/bin/docker pull vaultwarden/server:latest
ExecStartPre=-/usr/bin/docker stop vaultwarden
ExecStartPre=-/usr/bin/docker rm vaultwarden
ExecStart=/usr/bin/docker run \
  -p 8888:80 \
  -p 3012:3012 \
  --env-file /home/vaultwarden/etc/vaultwarden.conf \
  -v /home/vaultwarden/data/:/data/ \
  --rm --name vaultwarden vaultwarden/server:latest
ExecStopPost=-/usr/bin/docker rm vaultwarden
Restart=Always
RestartSec=30s
Type=notify
NotifyAccess=all

[Install]
WantedBy=multi-user.target
```

**Comment by BlackDex** [comment]
Well running this via systemd sounds a bit overkill, if docker is already started by systemd, then just use `docker run --restart-unless-stopped` or using docker-compose with the same kind of option.

**Reply by BlackDex** [reply]
I have removed that wiki page since it's not really useful, also since systemd-docker is no longer maintained for 6 or 7 years already.

**Comment by ctschach** [comment]
Hmmm....okay....

I'm quite good when it comes to system-level unix admin, but I never really spent much time with docker myself.

So what would be the right way to launch vaultwarden via docker with the whole set of parameters as shown in my code. Just want to make sure things are running at boot-time and restart if they crash...

Thanks...

**Reply by BlackDex** [reply]
Either use
```bash
docker run --name vaultwarden \
  -d --restart unless-stopped \
  -e DOMAIN=https://my.domain.tld \
  -v /vw-data/:/data/ \
  -p 80:80 \
  -p 3012:3012 \
  vaultwarden/server
```

Or use docker-compose:
```yml
---
version: '3.8'

services:
  vaultwarden:
    image: vaultwarden/server
    container_name: vaultwarden
    restart: unless-stopped
    volumes:
      - /vw-data/:/data/
    environment:
      - DOMAIN=https://my.domain.tld
    ports:
      - 80:80
      - 3012:3012
```

Add multiple `-e` or `environment` lines where needed to have your base config defined.
The `restart: unless-stopped` ensures the process starts again after a reboot if it wasn't shutdown before.
You can also set it to `always` if you always want to have it restarted.

**Comment by daydiff** [comment]
If you really want to run your container with systemd you should use Podman instead. It's a Linux-native tool, works perfectly with the systemd and compatible with docker containers. Moreover, it supports the same syntax as docker, so you don't need to learn it from scratch.

Then you run it:
```
podman run -d --name vaultwarden -v /vw-data/:/data/ -p 8000:80 vaultwarden/server:latest
```
Create a systemd config:
```
[Unit]
Description=Vaultwarden/Bitwarden Server (Rust Edition)
Documentation=https://github.com/dani-garcia/vaultwarden
Wants=syslog.service

[Service]
Restart=on-failure
ExecStart=/usr/bin/podman start -a vaultwarden
ExecStop=/usr/bin/podman stop vaultwarden

[Install]
WantedBy=multi-user.target
```

And reload the daemon
```
systemctl daemon-reload
```
For more details see this post: https://atabakoff.com/how-to-run-vaultwarden-in-podman-as-a-systemd-service/
